File: strategy/generator/pulse/wae.py
from dataclasses import dataclass

from core.models.parameter import CategoricalParameter, Parameter, StaticParameter
from core.models.smooth import Smooth
from core.models.source import SourceType

from .base import Pulse, PulseType


@dataclass(frozen=True)
class WaePulse(Pulse):
    type: PulseType = PulseType.Wae
    source: Parameter = StaticParameter(SourceType.CLOSE)
    smooth: Parameter = CategoricalParameter(Smooth)
    period_fast: Parameter = StaticParameter(10.0)
    period_slow: Parameter = StaticParameter(29.0)
    smooth_bb: Parameter = CategoricalParameter(Smooth)
    period_bb: Parameter = StaticParameter(13.0)
    factor: Parameter = StaticParameter(1.2)
    strength: Parameter = StaticParameter(69.0)

File: ta_lib/strategies/pulse/src/wae.rs
use base::prelude::*;
use core::prelude::*;
use timeseries::prelude::*;
use volatility::bb;

pub struct WaePulse {
    source: SourceType,
    smooth: Smooth,
    period_fast: usize,
    period_slow: usize,
    smooth_bb: Smooth,
    period_bb: usize,
    factor: f32,
    strength: f32,
}

impl WaePulse {
    pub fn new(
        source: SourceType,
        smooth: Smooth,
        period_fast: f32,
        period_slow: f32,
        smooth_bb: Smooth,
        period_bb: f32,
        factor: f32,
        strength: f32,
    ) -> Self {
        Self {
            source,
            smooth,
            period_fast: period_fast as usize,
            period_slow: period_slow as usize,
            smooth_bb,
            period_bb: period_bb as usize,
            factor,
            strength,
        }
    }
}

impl Pulse for WaePulse {
    fn lookback(&self) -> usize {
        let adj_lookback = std::cmp::max(self.period_fast, self.period_slow);
        std::cmp::max(adj_lookback, self.period_bb)
    }

    fn assess(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let source = data.source(self.source);

        let (upper_bb, _, lower_bb) = bb(&source, self.smooth_bb, self.period_bb, self.factor);

        let e = upper_bb - lower_bb;

        let diff =
            self.strength * source.spread_diff(self.smooth, self.period_fast, self.period_slow, 1);

        let zero = Series::zero(data.len());

        let up = iff!(diff.sgte(&ZERO), diff, zero);
        let down = iff!(diff.slt(&ZERO), diff.negate(), zero);

        (
            up.sgt(&up.shift(1)) & up.sgt(&e),
            down.sgt(&down.shift(1)) & down.sgt(&e),
        )
    }
}
